pyvrp/cpp/SubPopulation.cpp

```diff
@@ -0,0 +1,156 @@
+{
+}
+
+SubPopulation::~SubPopulation()
+{
+    for (auto &item : items)
+        delete item.individual;
+}
+
+void SubPopulation::add(Individual const *individual)
+{
+    // Copy the given individual into a new memory location, and use that from
+    // now on.
+    individual = new Individual(*individual);
+    Item item = {&params, individual, 0.0, {}};
+
+    for (auto &other : items)  // update distance to other individuals
+    {
+        auto const div = divOp(data, *individual, *other.individual);
+        auto cmp = [](auto &elem, auto &value) { return elem.first < value; };
+
+        auto &oProx = other.proximity;
+        auto place = std::lower_bound(oProx.begin(), oProx.end(), div, cmp);
+        oProx.emplace(place, div, individual);
+
+        auto &iProx = item.proximity;
+        place = std::lower_bound(iProx.begin(), iProx.end(), div, cmp);
+        iProx.emplace(place, div, other.individual);
+    }
+
+    items.push_back(item);
+    updateFitness();
+
+    if (size() > params.maxPopSize())
+        purge();
+}
+
+size_t SubPopulation::size() const { return items.size(); }
+
+SubPopulation::Item const &SubPopulation::operator[](size_t idx) const
+{
+    return items[idx];
+}
+
+std::vector<SubPopulation::Item>::const_iterator SubPopulation::cbegin() const
+{
+    return items.cbegin();
+}
+
+std::vector<SubPopulation::Item>::const_iterator SubPopulation::cend() const
+{
+    return items.cend();
+}
+
+void SubPopulation::remove(
+    std::vector<SubPopulation::Item>::iterator const &iterator)
+{
+    for (auto &[params, individual, fitness, proximity] : items)
+        // Remove individual from other proximities.
+        for (size_t idx = 0; idx != proximity.size(); ++idx)
+            if (proximity[idx].second == iterator->individual)
+            {
+                proximity.erase(proximity.begin() + idx);
+                break;
+            }
+
+    delete iterator->individual;  // dispose of manually allocated memory
+    items.erase(iterator);        // before the item is removed.
+}
+
+void SubPopulation::purge()
+{
+    while (size() > params.minPopSize)
+    {
+        // Remove duplicates from the subpopulation (if they exist)
+        auto const pred = [&](auto &iterator)
+        {
+            return !iterator.proximity.empty()
+                   && *iterator.proximity[0].second == *iterator.individual;
+        };
+
+        auto const duplicate = std::find_if(items.begin(), items.end(), pred);
+
+        if (duplicate == items.end())  // there are no more duplicates
+            break;
+
+        remove(duplicate);
+    }
+
+    while (size() > params.minPopSize)
+    {
+        // Remove the worst individual (worst in terms of biased fitness)
+        updateFitness();
+
+        auto const worstFitness = std::max_element(
+            items.begin(),
+            items.end(),
+            [](auto const &a, auto const &b) { return a.fitness < b.fitness; });
+
+        remove(worstFitness);
+    }
```

Comment:
Shouldn't we have an `updateFitness()` here after we remove the last individual? If this is indeed missing, I think we should add a test that fails for the current implementation.

Reply:
The current implementation does
![image](https://user-images.githubusercontent.com/16272507/221377355-f5b25b4e-84e7-4385-be65-b3406b8b2d15.png)
Which this function exactly replicates. We could move the call to update the biased fitness to the `remove` member altogether and make it a private method. I think that makes more sense. What do you think?

Reply:
I agree so the current implementation seems incorrect? Downside of moving it to remove (which is already private I think?) is that you do not need to update the biased fitness in between of removing multiple exact duplicates so maybe just keep it but add the one update? Or let remove do it by default but make it optional to not do it for efficiency? I don't care that much.

Reply:
I'm not really worried about performance here tbh. Everything on the C++ side is already (mostly) a rounding error; the whole subpopulation takes <1% of our current runtimes.

So I think we can make it simpler for ourselves and just always update after removing. That way we know it's correct without having to think about exceptional cases. We just update when adding and update when removing.

Reply:
>so the current implementation seems incorrect? 

So yes, I think so.

Reply:
Changing this hasn't impacted the individuals in the subpopulation when compared to the implementation on current main, so in that regard we're OK. I also ran once more with the profiler, and `Population::add` is still around 0.2s on 1000 iteration runs of `RC1_10_1`. So I think we're good here!